File: youtube-links/src/clients/UserApiClient.ts
import type { PagedList } from "@/shared/abstractions/PagedList";
import type { ConfirmEmail } from "@/shared/features/users/commands/ConfirmEmail";
import type { Login } from "@/shared/features/users/commands/Login";
import type { Register } from "@/shared/features/users/commands/Register";
import type { UpdateUserTheme } from "@/shared/features/users/commands/UpdateUserTheme";
import type { GetAllUsers } from "@/shared/features/users/queries/GetAllUsers";
import type { JwtDto } from "@/shared/features/users/responses/JwtDto";
import type { UserDto } from "@/shared/features/users/responses/UserDto";
import { apiClient, type IApiClient } from "./ApiClient";

export interface IUserApiClient {
    ConfirmEmail(command: ConfirmEmail.Command): Promise<boolean>;
    Login(command: Login.Command): Promise<JwtDto>;
    Register(command: Register.Command): Promise<void>;
    UpdateUserTheme(command: UpdateUserTheme.Command): Promise<void>;
    GetAllUsers(query: GetAllUsers.Query): Promise<PagedList<UserDto>>;
    GetUser(id: number): Promise<UserDto>;
}

export class UserApiClient implements IUserApiClient {
    private readonly apiClient: IApiClient;
    private readonly url: string = "api/users";

    constructor() {
        this.apiClient = apiClient;
    }

    async ConfirmEmail(command: ConfirmEmail.Command): Promise<boolean> {
        return await this.apiClient.post<ConfirmEmail.Command, boolean>(`${this.url}/confirmEmail`, command);
    }

    async Login(command: Login.Command): Promise<JwtDto> {
        return await this.apiClient.post<Login.Command, JwtDto>(`${this.url}/login`, command);
    }

    async Register(command: Register.Command): Promise<void> {
        return await this.apiClient.postWithoutResponse<Register.Command>(`${this.url}/register`, command);
    }

    async UpdateUserTheme(command: UpdateUserTheme.Command): Promise<void> {
        return await this.apiClient.put<UpdateUserTheme.Command>(`${this.url}/${command.id}/theme`, command);
    }

    async GetAllUsers(query: GetAllUsers.Query): Promise<PagedList<UserDto>> {
        return await this.apiClient.post<GetAllUsers.Query, PagedList<UserDto>>(`${this.url}/all`, query);
    }

    async GetUser(id: number): Promise<UserDto> {
        return await this.apiClient.getWithResponse<UserDto>(`${this.url}/${id}`);
    }
}

export const userApiClient = new UserApiClient();

File: YoutubeLinks.Blazor/Clients/UserApiClient.cs
using YoutubeLinks.Shared.Abstractions;
using YoutubeLinks.Shared.Features.Users.Commands;
using YoutubeLinks.Shared.Features.Users.Queries;
using YoutubeLinks.Shared.Features.Users.Responses;

namespace YoutubeLinks.Blazor.Clients;

public interface IUserApiClient
{
    Task<bool> ConfirmEmail(ConfirmEmail.Command command);
    Task ForgotPassword(ForgotPassword.Command command);
    Task<JwtDto> Login(Login.Command command);
    Task<JwtDto> RefreshToken(RefreshToken.Command command);
    Task Register(Register.Command command);
    Task ResendConfirmationEmail(ResendConfirmationEmail.Command command);
    Task<bool> ResetPassword(ResetPassword.Command command);
    Task UpdateUserTheme(UpdateUserTheme.Command command);
    Task<PagedList<UserDto>> GetAllUsers(GetAllUsers.Query query);
    Task<UserDto> GetUser(int id);
}

public class UserApiClient : IUserApiClient
{
    private const string Url = "api/users";
    private readonly IApiClient _apiClient;

    public UserApiClient(IApiClient apiClient)
    {
        _apiClient = apiClient;
    }

    public async Task<bool> ConfirmEmail(ConfirmEmail.Command command)
        => await _apiClient.Post<ConfirmEmail.Command, bool>($"{Url}/confirmEmail", command);

    public async Task ForgotPassword(ForgotPassword.Command command)
        => await _apiClient.Post($"{Url}/forgotPassword", command);

    public async Task<JwtDto> Login(Login.Command command)
        => await _apiClient.Post<Login.Command, JwtDto>($"{Url}/login", command);

    public async Task<JwtDto> RefreshToken(RefreshToken.Command command)
        => await _apiClient.Post<RefreshToken.Command, JwtDto>($"{Url}/refresh-token", command);

    public async Task Register(Register.Command command)
        => await _apiClient.Post($"{Url}/register", command);

    public async Task ResendConfirmationEmail(ResendConfirmationEmail.Command command)
        => await _apiClient.Post($"{Url}/resendConfirmationEmail", command);

    public async Task<bool> ResetPassword(ResetPassword.Command command)
        => await _apiClient.Post<ResetPassword.Command, bool>($"{Url}/resetPassword", command);

    public async Task UpdateUserTheme(UpdateUserTheme.Command command)
        => await _apiClient.Put($"{Url}/{command.Id}/theme", command);

    public async Task<PagedList<UserDto>> GetAllUsers(GetAllUsers.Query query)
        => await _apiClient.Post<GetAllUsers.Query, PagedList<UserDto>>($"{Url}/all", query);

    public async Task<UserDto> GetUser(int id)
        => await _apiClient.Get<UserDto>($"{Url}/{id}");
}